File: 2025-06-16/Valid_Palindrome_II.java
package patterns.two_pointers;

public class Valid_Palindrome_II {
    public static boolean helper(String s) {
        if(s == null || s.isEmpty()) return true;

        int left = 0, right = s.length()-1, remove = 1;
        while(left < right) {
            if(s.charAt(left) != s.charAt(right)) {
                if(remove-- == 0) return false;
                if(s.charAt(left) == s.charAt(right-1)) right--;
                else left++;
                continue;
            }
            left++;
            right--;
        }

        return true;
    }

    public static boolean validPalindrome(String s) {
        return helper(s) || helper(new StringBuilder(s).reverse().toString());
    }

    public static void main(String[] args) {
        System.out.println(validPalindrome("eceec"));
    }
}

File: 2026-01-25/Valid_Palindrome_II.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        if not s:
            return False

        def isValid(left, right):
            while left < right:
                if s[left] != s[right]:
                    return False

                left += 1
                right -= 1

            return True

        l, r = 0, len(s)-1
        while l < r:
            if s[l] != s[r]:
                return isValid(l, r-1) or isValid(l+1, r)

            l += 1
            r -= 1

        return True


sol = Solution()
print(sol.validPalindrome("daba"))